18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,106 @@
+import React, { useEffect, useState, useRef } from 'react';
+import './Chat.css';
+import Message from './components/Message';
+import io from 'socket.io-client';
+
+const ENDPOINT = "http://localhost:3001";
+const socket = io(ENDPOINT, {transports: ['websocket']});
+socket.connect();
+
+function Chat() {
+  const defaultName = 'Chel';
+  const [message, setMessage] = useState('');
+  const [listOfMessages, setListOfMessages] = useState([]);
+  const [username, setUsername] = useState(defaultName);
+  const [typing, setTyping] = useState(null);
+  const name = useRef(null);
+
+  function changeUserName() {
+    setUsername(name.current.value);
+    socket.emit("change name", name.current.value);
+  }
+
+  useEffect(() => {
+    socket.emit("add user", defaultName);
+    return () => {
+      socket.disconnect();
+    }
+  }, []);
+
+  socket.on('reg as user', (data) => {
+    setUsername(data);
+  });
+
+  socket.on('new message', (data) => {
+    setListOfMessages([...listOfMessages, data]);
+  });
+
+  socket.on('typing', (data) => {
+    setTimeout(() => {
+      setTyping(null);
+    }, 1000)
+    setTyping(data);
+  });
+
+  function type(e) {
+    setMessage(e.target.value);
+    socket.emit("typing", username);
```

Comment:
if the user is a super typer man, you will get a huge number of requests from him. A good practice is to use the debounce method for calling function no more than one time per 2-4 secs.
Unfortunately, javascript doesn't have this method. You can write it by yourself or use some library.

Reply:
i've added debounce method and multiple typing, but i think there is a problem with my timeouts
when im trying run 3 windows with react app, typing with one locking another 2 browser windows :D

Reply:
let's use library like that: https://www.npmjs.com/package/debounce